cmake/CeleritasUtils.cmake

```diff
@@ -268,6 +283,36 @@ endfunction()
 
 #-----------------------------------------------------------------------------#
 
+macro(celeritas_find_or_external_package package)
+  set(_var "CELERITAS_EXTERNAL_${package}")
+  set(_use_external ${_var})
+  if(NOT ${_use_external})
+    if(NOT "${${package}_FOUND}")
+      # Only look for the package if it's not "found" (which can occur if
+      # Celeritas is a subproject)
+      find_package(${package} ${ARGN})
+    endif()
+    if(NOT DEFINED ${_var})
+      set(_found "${${package}_FOUND}")
+      if(NOT _found)
+        set(_use_external ON)
+        message(STATUS "Set external ${package} to fetch and build")
+      endif()
+      set("${_var}" "${_use_external}" CACHE BOOL
+        "Fetch and build ${package}")
+      mark_as_advanced(${_var})
+      unset(_found)
+    endif()
+  endif()
+  if(_use_external)
+    set(CELERITAS_EXTERNAL TRUE)
+  endif()
+  unset(_use_external)
+  unset(_var)
+endmacro()
```

Comment:
Unless I missed some details, this could be replaced by:
```cmake
FetchContent_Declare(
package
FIND_PACKAGE_ARGS
)
```
which also tries ```find_package``` before downloading the dependency.

Reply:
Hmm good point, I didn't see that when I first started looking at `FetchContent`. I suppose the difference is that if we do it that way, the content fetching will have to go in the top level of our cmakelists (since `find_package` declares "local" variables). It's nice to be able to scope the external packages inside of a subdirectory so you can build `external/all` and  define local variables to override those projects.

It also looks like this is new to 3.24 which is *very* new: I'd like to keep 3 years of compatibility for dependencies if possible, which means ~3.20 is the newest we can do.

Reply:
make sense, if we want to support 3.20 let's keep the custom find_or_fetch